### PR TITLE
Feature: patient registration report

### LINKED ISSUE #470: Implement Patient Registration PDF report
In 1.X, we have a comprehensive patient report, that shows patient registrations by time period.  This report is rendered in HTML, and may crash the browser if the data to be rendered is too large.  To avoid this, I propose that we implement a PDF report in 2.X that is rendered on the server.

**Implementation Details**
A print button should be available at the top right of the navigation on the Patient Registry (`/patients`) page.  This button should print whatever is currently matching in the grid.  A modal should pop up with the embedded PDF, allowing the user to download it, open it in a new window, or cancel.

### PR DESCRIPTION
**Pull request by jniles**
This pull request implements the patient registration report as a PDF report on the client.  It is opened by default via a "print" button on the patient registry page.  The report is available as JSON, HTML, and PDF using the renderers implemented by @sfount.

**Screenshots**
html
_[Fig 1]: HTML Report_
pdf
_[Fig 2]: PDF Report_

This is a very rough PR, and should not be considered final.  Here are some TODOs:
1. Filtering on the client should be additive, and use a key:value model, so that each filter takes in one parameter.  For example, `name:jniles` would be stored in the JSON `{ name : 'jniles' }`, and  be sent to the server as `?name=jniles`.
2. The Receipt Service is now overloaded with reports.  We need to cleanly separate Receipts as a subset of reports, and ensure that the modal and controllers are appropriately named.

These should be tackled in a subsequent PR.

Closes #470.

---

Thank you for contributing!

Before submitting this pull request, please verify that you have:
- [x] Run your code through JSHint.  [Check out our styleguide](https://github.com/IMA-WorldHealth/bhima-2.X/blob/master/docs/STYLEGUIDE.md).
- [x] Run integration tests.
- [x] Run end-to-end tests.
- [x] Accurately described the changes your are making in this pull request.

For a more detailed checklist, [see the official review checklist](https://docs.google.com/document/d/1nupLVLRXgSZJQo_acLgrwvPnN8RukfSiwRhSToj81uU/pub) that this PR will be evaluated against.

Ensuring that the above checkboxes are completed will help speed the review process and help build a stronger application.  Thanks!

### CODE REVIEW

**Comment by jniles** [comment]
@sfount, in the spirit of living in the moment, this PR was made a moment ago, and the review should only take a moment.  Could I get one?

File: client/src/js/services/receipts/ReceiptModal.js [hunk 1]
```diff
@@ -44,7 +45,7 @@ function ReceiptModal(Modal, Receipts) {
       notifyCreated : notifyCreated
     };
 
-    var invoiceRequest = Receipts.invoice(uuid, { render : options.renderer });
+    var invoiceRequest = Receipts.invoice(uuid, { renderer : options.renderer });
```
Comment (sfount) [inline client/src/js/services/receipts/ReceiptModal.js:48]: :+1:

File: client/src/js/services/receipts/ReceiptService.js [hunk 2]
```diff
@@ -18,39 +18,54 @@ function ReceiptService($http, util) {
...
-  
+
   function patient(uuid, options) {
     var route ='/reports/patient/'.concat(uuid);
     var responseType = null;
-    
-    if (options.render === renderers.PDF) { 
+
+    if (options.renderer === renderers.PDF) {
+      responseType = 'arraybuffer';
```
Comment (sfount) [inline client/src/js/services/receipts/ReceiptService.js:51]: Just a note to say that in the future we could wrap this funcitonality in a utility method - there are a lot of crucial constants to get wrong here. 

First thing that comes to mind:

``` js
let params;
if (options.renderer === renderers.PDF) { 
  // returns 
  // { 
  //   params : options,
  //   responseType : 'arraybuffer'
  //  }
  params = utility.configurePDFParams(options);
}

$http.get(route, params);
```

What do you think?
@IMA-WorldHealth/local-contributors

Comment (sfount) [inline client/src/js/services/receipts/ReceiptService.js]: Just a note to say that in the future we could wrap this funcitonality in a utility method - there are a lot of crucial constants to get wrong here. 

First thing that comes to mind:
```js
let params;
if (options.renderer === renderers.PDF) { 
  // returns 
  // { 
  //   params : options,
  //   responseType : 'arraybuffer'
  //  }
  params = utility.configurePDFParams(options);
}

$http.get(route, params);
```

File: server/controllers/medical/patients/index.js [hunk 3]
```diff
@@ -284,137 +286,82 @@ function hospitalNumberExists(req, res, next) {
...
-    LIMIT 10;`;
-
-  // man. That's a lot of matches
-  db.exec(sql, [match, match, match, match, match, match, match, match, match, match, match, match])
-  .then(function (rows) {
-    res.status(200).json(rows);
-  })
-  .catch(next)
-  .done();
+ * @function logVisit
```
Comment (sfount) [inline server/controllers/medical/patients/index.js]: All of this was removed in https://github.com/IMA-WorldHealth/bhima-2.X/pull/477

It must have been reintroduced by accident in a rebase - can this be removed?

EDIT 
- That's anything to do with logVisit / visit

File: server/controllers/medical/reports/patient.receipt.js [hunk 4]
```diff
@@ -10,36 +10,36 @@ supportedRender.html = require('../../../lib/renderers/html');
...
 exports.build = build;
 
-function build(req, res, next) {  
+function build(req, res, next) {
   var queryString = req.query;
   var patientID = req.params.uuid;
-  
-  var renderTarget = queryString.render || defaultRender;
+
+  var renderTarget = queryString.renderer || defaultRender;
```
Comment (sfount) [inline server/controllers/medical/reports/patient.receipt.js:28]: :+1:

**Comment by sfount** [comment]
@jniles This has been reviewed. 

If you could remove any code re-introducing logVisit/ visiit on the server patient controller this can be merged anytime!

File: client/src/js/services/receipts/ReceiptService.js (hunk 2 above)
Comment (jniles) [inline client/src/js/services/receipts/ReceiptService.js:51]: Sounds good to me.

File: server/controllers/medical/patients/index.js (hunk 3 above)
Comment (jniles) [inline server/controllers/medical/patients/index.js]: Gotcha.  I was a bit confused in the changelog about what happened and favored preserving old code to removing it.  I'll remove it now.

**Comment by jniles** [comment]
@sfount changes have landed.

**Comment by jniles** [comment]
@sfount, rebased.

**Pull request merged**